packages/effect/src/internal/stream.ts

```diff
@@ -4686,6 +4686,39 @@ export const range = (min: number, max: number, chunkSize = DefaultChunkSize): S
     return new StreamImpl(go(min, max, chunkSize))
   })
 
+export const raceAll = <T extends ReadonlyArray<Stream.Stream<any, any, any>>>(
+  ...streams: T
+): Stream.Stream<
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<infer A, any, any> ? A : never,
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<any, infer E, any> ? E : never,
+  [T[number]] extends [never] ? never : T[number] extends Stream.Stream<any, any, infer R> ? R : never
+> => {
+  let finished = false
+  return unwrap(
+    Effect.gen(function*() {
+      const halts = yield* Effect.all(streams.map(() => Deferred.make<void>()))
```

Comment:
You should be able to do this with only one Deferred, and catching the signal for the elected stream.